rules/xcodeproj.bzl

```diff
@@ -0,0 +1,263 @@
+    if AppleBundleInfo in target:
+        bundle_info = target[AppleBundleInfo]
+        srcs = []
+        bazel_name = target.label.name
+        info = struct(
+            name = bundle_info.bundle_name,
+            bundle_extension = bundle_info.bundle_extension,
+            package = target.label.package,
+            bazel_name = bazel_name,
+            srcs = depset(srcs, transitive = _get_transitive(deps, _SrcsInfo, "srcs")),
+            build_files = depset([ctx.build_file_path], transitive = _get_transitive(deps, _SrcsInfo, "build_files")),
+            product_type = bundle_info.product_type[len("com.apple.product-type."):],
+        )
+        providers.append(
+            _SrcsInfo(
+                srcs = info.srcs,
+                build_files = depset([ctx.build_file_path]),
+                direct_srcs = srcs,
+            ),
+        )
+        target_info = _TargetInfo(target = info, targets = depset([info], transitive = _get_transitive(deps, _TargetInfo, "targets")))
+        providers.append(target_info)
+    elif ctx.rule.kind == "apple_framework_packaging":
+        srcs = []
+        info = struct(
+            name = target.label.name,
+            package = target.label.package,
+            bazel_name = target.label.name,
+            srcs = depset(srcs, transitive = _get_transitive(deps, _SrcsInfo, "srcs")),
+            build_files = depset([ctx.build_file_path], transitive = _get_transitive(deps, _SrcsInfo, "build_files")),
+            product_type = "framework",
+        )
+        target_info = _TargetInfo(target = info, targets = depset([info], transitive = _get_transitive(deps, _TargetInfo, "targets")))
+        providers.append(target_info)
+    else:
+        srcs = []
+        for attr in _dir(ctx.rule.files):
+            srcs += getattr(ctx.rule.files, attr, [])
+        srcs = [f for f in srcs if not f.path.startswith("external/") and f.is_source]
+
+        providers.append(
+            _SrcsInfo(
+                srcs = depset(srcs, transitive = _get_transitive(deps, _SrcsInfo, "srcs")),
+                build_files = depset([ctx.build_file_path], transitive = _get_transitive(deps, _SrcsInfo, "build_files")),
+                direct_srcs = srcs,
+            ),
+        )
+
+        info = None
+        actual = None
+        if ctx.rule.kind in ("test_suite"):
+            actual = getattr(ctx.rule.attr, "tests")[0]
+        elif ctx.rule.kind in ("alias"):
+            actual = getattr(ctx.rule.attr, "actual")
+        if actual and _TargetInfo in actual:
+            info = actual[_TargetInfo].target
+
+        providers.append(
+            _TargetInfo(target = info, targets = depset(transitive = _get_transitive(deps, _TargetInfo, "targets"))),
+        )
+
+    return providers
+
+_xcodeproj_aspect = aspect(
+    implementation = _xcodeproj_aspect_impl,
+    attr_aspects = ["deps", "actual", "tests", "infoplists", "entitlements", "resources", "test_host"],
+)
+
+def _xcodeproj_impl(ctx):
```

Comment:
For these yaml files I wonder if it wouldn't be better to have a template and generate an action that expands the template... it seems more maintainable than concatenating strings...

Reply:
Well switch to using struct to_json eventually